cmd/app/serve.go

```diff
@@ -314,3 +346,79 @@ func checkServeCmdConfigFile() error {
 	}
 	return nil
 }
+
+func StartDuplexServer(ctx context.Context, cfg *config.FulcioConfig, ctClient *ctclient.LogClient, baseca ca.CertificateAuthority, port, metricsPort int) error {
+	logger, opts := log.SetupGRPCLogging()
+
+	d := duplex.New(
+		port,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		runtime.WithMetadata(extractOIDCTokenFromAuthHeader),
+		grpc.UnaryInterceptor(grpcmw.ChainUnaryServer(
+			grpc_recovery.UnaryServerInterceptor(grpc_recovery.WithRecoveryHandlerContext(panicRecoveryHandler)), // recovers from per-transaction panics elegantly, so put it first
+			middleware.UnaryRequestID(middleware.UseXRequestIDMetadataOption(true), middleware.XRequestMetadataLimitOption(128)),
+			grpc_zap.UnaryServerInterceptor(logger, opts...),
+			PassFulcioConfigThruContext(cfg),
+			grpc_prometheus.UnaryServerInterceptor,
+		)),
+		grpc.MaxRecvMsgSize(int(maxMsgSize)),
+		runtime.WithForwardResponseOption(HTTPResponseModifier),
+	)
+
+	// GRPC server
+	grpcCAServer := server.NewGRPCCAServer(ctClient, baseca)
+	protobuf.RegisterCAServer(d.Server, grpcCAServer)
+	if err := d.RegisterHandler(ctx, protobuf.RegisterCAHandlerFromEndpoint); err != nil {
+		return fmt.Errorf("registering grpc ca handler: %w", err)
+	}
+
+	// Legacy server
+	legacyGRPCCAServer := server.NewLegacyGRPCCAServer(grpcCAServer)
+	legacy.RegisterCAServer(d.Server, legacyGRPCCAServer)
+	if err := d.RegisterHandler(ctx, legacy.RegisterCAHandlerFromEndpoint); err != nil {
+		return fmt.Errorf("registering legacy grpc ca handler: %w", err)
+	}
+
+	// Prometheus
+	reg := prometheus.NewRegistry()
+	grpcMetrics := grpc_prometheus.DefaultServerMetrics
+	grpcMetrics.EnableHandlingTimeHistogram()
+	reg.MustRegister(grpcMetrics, server.MetricLatency, server.RequestsCount)
+	grpc_prometheus.Register(d.Server)
+
+	// Register prometheus handle.
+	d.RegisterListenAndServeMetrics(metricsPort, false)
+
+	logger.Info("Starting duplex server...")
+	if err := d.ListenAndServe(ctx); err != nil {
+		return fmt.Errorf("duplex server: %w", err)
+	}
+	return nil
+}
+
+// The fulcio HTTP legacy client requires a 201 response status code to pass
+// However, even though the legacy client sets the 201 code, GRPC will automatically
+// change it to 200, causing the cert request to fail
+// GRPC recommends controlling HTTP status codes with this response modifier
+// https://grpc-ecosystem.github.io/grpc-gateway/docs/mapping/customizing_your_gateway/#controlling-http-response-status-codes
+// This is required to use fulcio with duplex.
+func HTTPResponseModifier(ctx context.Context, w http.ResponseWriter, p proto.Message) error {
```

Comment:
could we re-use https://github.com/sigstore/fulcio/blob/bc08c5b641033bcd9f604e202c84c8c9fde502dd/cmd/app/http.go#L97 instead of this implementation? The existing one deals with returning the SCT in a response header (which was part of v1 IIUC), and also strips out all GRPC response headers (to hide the internal implementation detail).

Reply:
Thanks! I hadn't noticed that before, it should work as well. I've updated it to use this one!